vms/platformvm/state/state.go

```diff
@@ -1576,24 +1602,58 @@ func (s *state) GetStatelessBlock(blockID ids.ID) (blocks.Block, error) {
 		return nil, err
 	}
 
-	// Note: stored blocks are verified, so it's safe to unmarshal them with GenesisCodec
-	blkState := stateBlk{}
-	if _, err := blocks.GenesisCodec.Unmarshal(blkBytes, &blkState); err != nil {
+	blk, status, _, err := parseStoredBlock(blkBytes)
+	if err != nil {
 		return nil, err
 	}
 
-	if blkState.Status != choices.Accepted {
+	if status != choices.Accepted {
 		s.blockCache.Put(blockID, nil)
 		return nil, database.ErrNotFound
 	}
 
-	blkState.Blk, err = blocks.Parse(blocks.GenesisCodec, blkState.Bytes)
+	s.blockCache.Put(blockID, blk)
+	return blk, nil
+}
+
+func (s *state) writeBlockIDs() error {
+	for height, blkID := range s.addedBlockIDs {
+		heightKey := database.PackUInt64(height)
+
+		delete(s.addedBlockIDs, height)
+		s.blockIDCache.Put(height, blkID)
+		if err := database.PutID(s.blockIDDB, heightKey, blkID); err != nil {
+			return fmt.Errorf("failed to add blockID: %w", err)
+		}
+	}
+	return nil
+}
```

Comment:
I'm wondering if this would be more clear if we combined `writeBlocks` and `writeBlockIDs`. They are different indices... But they are tightly coupled

Reply:
Did a pass on this - not sure if I like it quite yet